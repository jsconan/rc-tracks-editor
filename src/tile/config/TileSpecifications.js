/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gives the specifications for the track tiles with respect to the initial constraints.
 */
export class TileSpecifications {
    /**
     * Defines the specifications for the track tiles from the given constraints.
     * @param {object} config - A set of config options.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @param {boolean} unlockRatio - Unlock the maximum size ratio some tiles are applying.
     */

    constructor({ laneWidth = 20, barrierWidth = 1, barrierChunks = 4, maxRatio = 4, unlockRatio = false } = {}) {
        this.setLaneWidth(laneWidth);
        this.setBarrierWidth(barrierWidth);
        this.setBarrierChunks(barrierChunks);
        this.setMaxRatio(maxRatio);
        this.setUnlockRatio(unlockRatio);
    }

    /**
     * The context identifier to retrieve the config from the context.
     * @type {string}
     */
    get contextId() {
        return this.constructor.CONTEXT_ID;
    }

    /**
     * The length of a tile with respect to the initial constraints.
     * @type {number}
     */
    get length() {
        return this.laneWidth * 1.25 + this.barrierWidth * 2;
    }

    /**
     * The width of a tile with respect to the initial constraints.
     * @type {number}
     */
    get width() {
        return this.laneWidth + this.barrierWidth * 2;
    }

    /**
     * The length of a barrier chunk with respect to the initial constraints.
     * @type {number}
     */
    get barrierLength() {
        return this.length / this.barrierChunks;
    }

    /**
     * The padding around the width of the tile to align with its length.
     * Note: tiles must fit in a square so that they can be flipped and rotated indifferently.
     * @type {number}
     */
    get padding() {
        return this.laneWidth / 8;
    }

    /**
     * Sets the width of the track lane (the distance between the barriers).
     * @param {number} laneWidth - The width of the track lane.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setLaneWidth(laneWidth) {
        this.laneWidth = Math.abs(laneWidth);

        return this;
    }

    /**
     * Sets the width of the barriers.
     * @param {number} barrierWidth - The width of the barriers.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierWidth(barrierWidth) {
        this.barrierWidth = Math.abs(barrierWidth);

        return this;
    }

    /**
     * Sets the number of barrier chunks per tile.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierChunks(barrierChunks) {
        this.barrierChunks = Math.abs(Math.round(barrierChunks) || 1);

        return this;
    }

    /**
     * Sets the maximum value for size ratios.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setMaxRatio(maxRatio) {
        this.maxRatio = Math.abs(Math.round(maxRatio) || 1);

        return this;
    }

    /**
     * Locks or unlocks the maximum size ratio some tiles are applying.
     * @param {boolean} unlockRatio - Whether or not unlock the size ratio.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setUnlockRatio(unlockRatio) {
        this.unlockRatio = !!unlockRatio;

        return this;
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The specifications object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * The context identifier to retrieve the config from the context.
 * @constant {string} TileSpecifications.CONTEXT_ID
 */
Object.defineProperty(TileSpecifications, 'CONTEXT_ID', {
    value: 'tileSpecifications',
    writable: false,
    enumerable: true,
    configurable: true
});
