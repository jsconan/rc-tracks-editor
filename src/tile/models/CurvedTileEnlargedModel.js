/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { quadrantRange, CURVED_TILE_ENLARGED_TYPE, TILE_DIRECTION_LEFT } from '../helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a track tile for an enlarged curve.
 */
export class CurvedTileEnlargedModel extends TileModel {
    /**
     * The minimum allowed size ratio for the tile.
     * @type {number}
     */
    get minRatio() {
        return 1;
    }

    /**
     * Computes the angle for rotating the tile to the expected direction.
     * @returns {number}
     */
    getDirectionAngle() {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.RIGHT_ANGLE;
        }

        return 0;
    }

    /**
     * Computes the angle of the curve with respect to the ratio.
     * @returns {number}
     */
    getCurveAngle() {
        return Vector2D.RIGHT_ANGLE;
    }

    /**
     * Computes the length of the curve side if any.
     * @returns {number}
     */
    getCurveSide() {
        return this.length / 2;
    }

    /**
     * Computes the number of barrier chunks on each side with respect to the ratio.
     * @returns {number}
     */
    getSideBarrierChunks() {
        return (this.specs.barrierChunks * this.ratio) / 2;
    }

    /**
     * Computes the number of barrier chunks for the inner curve with respect to the ratio.
     * @returns {number}
     */
    getInnerBarrierChunks() {
        if (this.ratio <= 1) {
            return this.specs.barrierChunks / 2;
        }
        return this.specs.barrierChunks * this.ratio;
    }

    /**
     * Computes the number of barrier chunks for an outer curve with respect to the ratio.
     * @returns {number}
     */
    getOuterBarrierChunks() {
        return (this.specs.barrierChunks / 2) * this.ratio;
    }

    /**
     * Computes the coordinates of the center point.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getCenterCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);
        const center = start.addScalarY(this.specs.length * (this.ratio - 0.5));

        return center.rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;
        const curveAngle = this.getCurveAngle();

        let center;
        if (this.direction === TILE_DIRECTION_LEFT) {
            center = start.addScalarX(radius);
        } else {
            center = start.subScalarX(radius);
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the angle of the output point with respect to the tile direction.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngle(angle = 0) {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.degrees(angle - Vector2D.RIGHT_ANGLE);
        }

        return Vector2D.degrees(angle + Vector2D.RIGHT_ANGLE);
    }

    /**
     * Computes the coordinates of the edge points with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D[]} - Returns a list of edge points.
     */
    getEdgesCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);
        const width = this.specs.width;
        const side = this.getCurveSide();
        const innerRadius = this.getInnerRadius();
        const roundRadius = this.getOuterRadius();
        const outerRadius = innerRadius + width;
        const centerRadius = innerRadius + width / 2;

        let startAngle, center, curveCenter;
        if (this.direction === TILE_DIRECTION_LEFT) {
            center = start.addScalarX(centerRadius);
            curveCenter = center.addCoord(-side, side);
            startAngle = Vector2D.RIGHT_ANGLE;
        } else {
            center = start.subScalarX(centerRadius);
            curveCenter = center.addScalar(side);
            startAngle = 0;
        }
        const endAngle = startAngle + Vector2D.RIGHT_ANGLE;

        const p0 = Vector2D.polar(innerRadius, startAngle, center).rotateAround(angle, start);
        const p1 = Vector2D.polar(outerRadius, startAngle, center).rotateAround(angle, start);
        const p2 = Vector2D.polar(roundRadius, startAngle, curveCenter).rotateAround(angle, start);
        const p3 = Vector2D.polar(roundRadius, endAngle, curveCenter).rotateAround(angle, start);
        const p4 = Vector2D.polar(outerRadius, endAngle, center).rotateAround(angle, start);
        const p5 = Vector2D.polar(innerRadius, endAngle, center).rotateAround(angle, start);

        const edges = [p0, p1, p2];

        const rotatedCenter = curveCenter.rotateAround(angle, start);
        const rotatedStartAngle = Vector2D.degrees(p2.sub(rotatedCenter).angle());
        const rotatedEndAngle = rotatedStartAngle + Vector2D.RIGHT_ANGLE;

        const edgeAngle = quadrantRange(rotatedStartAngle, rotatedEndAngle);
        if (edgeAngle !== null) {
            edges.push(Vector2D.polar(roundRadius, edgeAngle, rotatedCenter));
        }

        edges.push(p3, p4, p5);

        return edges;
    }
}

/**
 * The type of tile.
 * @constant {string} CurvedTileEnlargedModel.TYPE
 */
Object.defineProperty(CurvedTileEnlargedModel, 'TYPE', {
    value: CURVED_TILE_ENLARGED_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
