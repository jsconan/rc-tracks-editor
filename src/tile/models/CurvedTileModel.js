/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_TYPE, TILE_DIRECTION_LEFT } from '../helpers';
import { degrees, quadrantRange, RIGHT_ANGLE, STRAIGHT_ANGLE } from '../../core/helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a track tile for a curve.
 */
export class CurvedTileModel extends TileModel {
    /**
     * The actual length of the tile with respect to the ratio.
     * @type {number}
     */
    get length() {
        return this.specs.length;
    }

    /**
     * The actual width of the tile with respect to the ratio.
     * @type {number}
     */
    get width() {
        return this.specs.width;
    }

    /**
     * The minimum allowed size ratio for the tile.
     * @type {number}
     */
    get minRatio() {
        return 2 / this.specs.barrierChunks;
    }

    /**
     * Computes the angle for rotating the tile to the expected direction.
     * @returns {number}
     */
    getDirectionAngle() {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return RIGHT_ANGLE + (RIGHT_ANGLE / this.ratio) * (Math.max(1, this.ratio) - 1);
        }

        return 0;
    }

    /**
     * Computes the angle of the curve with respect to the ratio.
     * @returns {number}
     */
    getCurveAngle() {
        if (this.ratio < 1) {
            return RIGHT_ANGLE * this.ratio;
        }

        return RIGHT_ANGLE / this.ratio;
    }

    /**
     * Computes the coordinates of the center point.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getCenterCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const curveAngle = this.getCurveAngle();
        const curveCenter = this.getCurveCenter().add(start);
        const radius = this.getInnerRadius() + this.specs.width / 2;

        const p1 = Vector2D.polar(radius, 0, curveCenter);
        const p2 = p1.addScalarY(10);
        const p3 = Vector2D.polar(radius, curveAngle, curveCenter);
        const p4 = p3.add(Vector2D.polar(10, curveAngle + RIGHT_ANGLE));

        const center = Vector2D.intersect(p1, p2, p3, p4);

        return center.rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;

        let curveAngle, center;
        if (this.direction === TILE_DIRECTION_LEFT) {
            curveAngle = STRAIGHT_ANGLE - this.getCurveAngle();
            center = start.addScalarX(radius);
        } else {
            curveAngle = this.getCurveAngle();
            center = start.subScalarX(radius);
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the angle of the output point with respect to the tile direction.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngle(angle = 0) {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return degrees(angle - this.getCurveAngle());
        }

        return degrees(angle + this.getCurveAngle());
    }

    /**
     * Computes the coordinates of the edge points with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D[]} - Returns a list of edge points.
     */
    getEdgesCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);
        const width = this.specs.width;
        const innerRadius = this.getInnerRadius();
        const outerRadius = innerRadius + width;
        const centerRadius = innerRadius + width / 2;

        let startAngle, center;
        if (this.direction === TILE_DIRECTION_LEFT) {
            center = start.addScalarX(centerRadius);
            startAngle = STRAIGHT_ANGLE - this.getCurveAngle();
        } else {
            center = start.subScalarX(centerRadius);
            startAngle = 0;
        }
        const endAngle = startAngle + this.getCurveAngle();

        const p0 = Vector2D.polar(innerRadius, startAngle, center).rotateAround(angle, start);
        const p1 = Vector2D.polar(outerRadius, startAngle, center).rotateAround(angle, start);
        const p2 = Vector2D.polar(outerRadius, endAngle, center).rotateAround(angle, start);
        const p3 = Vector2D.polar(innerRadius, endAngle, center).rotateAround(angle, start);

        const edges = [p0, p1];

        const rotatedCenter = center.rotateAround(angle, start);
        const rotatedStartAngle = degrees(p1.sub(rotatedCenter).angle());
        const rotatedEndAngle = rotatedStartAngle + this.getCurveAngle();

        const edgeAngle = quadrantRange(rotatedStartAngle, rotatedEndAngle);
        if (edgeAngle !== null) {
            edges.push(Vector2D.polar(outerRadius, edgeAngle, rotatedCenter));
        }

        edges.push(p2, p3);

        return edges;
    }

    /**
     * Computes the parameters for rendering tile shapes at the expected position.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @returns {object} - Returns a set of parameters for rendering the tile shapes.
     */
    getShapeParameters(x = 0, y = 0) {
        const barrierWidth = this.specs.barrierWidth;

        const innerRadius = this.getInnerRadius();
        const outerRadius = this.getOuterRadius() - barrierWidth;
        const innerChunks = this.getInnerBarrierChunks();
        const outerChunks = this.getOuterBarrierChunks();
        const curveAngle = this.getCurveAngle();
        const curveCenter = this.getCurveCenter(x, y);

        const innerX = curveCenter.x + innerRadius;
        const innerY = curveCenter.y;
        const outerX = curveCenter.x + outerRadius;
        const outerY = curveCenter.y;

        const ground = {
            cx: curveCenter.x,
            cy: curveCenter.y,
            width: this.specs.width,
            radius: innerRadius,
            angle: curveAngle,
            start: 0
        };
        const innerBarrier = {
            chunks: innerChunks,
            width: barrierWidth,
            radius: innerRadius,
            angle: curveAngle,
            left: innerX,
            top: innerY,
            shift: 0
        };
        const outerBarrier = {
            chunks: outerChunks,
            width: barrierWidth,
            radius: outerRadius,
            angle: curveAngle,
            left: outerX,
            top: outerY,
            shift: 1
        };

        return {
            ground,
            innerBarrier,
            outerBarrier
        };
    }
}

/**
 * The type of tile.
 * @constant {string} CurvedTileModel.TYPE
 */
Object.defineProperty(CurvedTileModel, 'TYPE', {
    value: CURVED_TILE_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
