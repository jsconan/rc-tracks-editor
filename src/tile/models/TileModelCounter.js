/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Counter } from '../../core/models';
import { TILE_DIRECTION_RIGHT } from '../helpers';

/**
 * Extracts the tile model from a tile.
 * @param {TileModel} tile - The tile from which get the model.
 * @returns {TileModel} - The base model of the given tile.
 * @private
 */
const getModel = tile => tile.clone().setDirection(TILE_DIRECTION_RIGHT);

/**
 * Represents an observable counter for tile models.
 */
export class TileModelCounter extends Counter {
    /**
     * Creates an observable counter for tile models.
     * @param {TileModel[]|TileList} source - A list of tile model to initialize the counters.
     */
    constructor(source = null) {
        super();

        this.models = new Map();

        if (source) {
            this.load(source);
            source = void 0;
        }
    }

    /**
     * Exports the counters to an array.
     * @returns {tileCounter[]} - A list of tile counters.
     */
    getCounterList() {
        const counters = [];

        for (const [modelId, count] of this.entries()) {
            const model = this.getModel(modelId);
            counters.push({ modelId, model, count });
        }

        return counters;
    }

    /**
     * Gets the list of counted tile models for which the count if greater than 0.
     * @returns {TileModel[]} - Returns the list of counted tile models.
     */
    getModelList() {
        return [...this.models.values()];
    }

    /**
     * Gets the tile model registered under the given key.
     * @param {string} key - The key of the counter to get the tile model for.
     * @returns {TileModel} - The tile model registered under the given key.
     */
    getModel(key) {
        return this.models.get(key) || null;
    }

    /**
     * Increments the counter for the given tile model.
     * If the tile model was not yet listed, it will be added.
     * @param {TileModel} tile - The tile to count.
     * @returns {TileModelCounter} - Chains the instance.
     * @fires set
     * @fires addmodel
     * @fires addtile
     */
    add(tile) {
        const { modelId } = tile;

        this.increment(modelId, 1);

        if (!this.models.has(modelId)) {
            const model = getModel(tile);
            this.models.set(modelId, model);

            this.emit('addmodel', modelId, model);
        }

        this.emit('addtile', tile);

        return this;
    }

    /**
     * Decrements the counter for the given tile model.
     * If the counter reaches 0, it will be removed together with the related tile model.
     * @param {TileModel} tile - The tile to count.
     * @returns {TileModelCounter} - Chains the instance.
     * @fires set
     * @fires removemodel
     * @fires delete
     * @fires removetile
     */
    remove(tile) {
        const { modelId } = tile;
        const count = this.get(modelId);

        if (count > 0) {
            if (count > 1) {
                this.decrement(modelId, 1);
            } else {
                this.delete(modelId);
            }

            this.emit('removetile', tile);
        }

        return this;
    }

    /**
     * Removes a counter together with its related tile model.
     * @param {string} key - The key of the counter to delete.
     * @returns {boolean} - Returns `true` if a counter was deleted.
     * @fires removemodel
     * @fires delete
     */
    delete(key) {
        const model = this.models.get(key);
        const deleted = this.models.delete(key);

        if (deleted) {
            this.emit('removemodel', key, model);
        }

        return super.delete(key);
    }

    /**
     * Removes all counters, including the related tile models.
     * @returns {TileModelCounter} - Chains the instance.
     * @fires clear
     */
    clear() {
        this.models.clear();

        return super.clear();
    }

    /**
     * Loads counter from a list of tiles. The counters are cleared before, including the related tile models.
     * @param {*} iterator - An iterable object that can be used to set the counters.
     * @returns {TileModelCounter} - Chains the instance.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        this.models.clear();

        const dataIterator = iterator[Symbol.iterator]();
        const loadIterator = {
            next: () => {
                const next = dataIterator.next();

                if (next.done) {
                    return next;
                }

                const tile = next.value;
                const { modelId } = tile;

                if (!this.models.has(modelId)) {
                    this.models.set(modelId, getModel(tile));
                }

                const count = this.get(modelId) + 1;
                return { done: false, value: [modelId, count] };
            },

            [Symbol.iterator]() {
                return this;
            }
        };

        return super.load(loadIterator);
    }
}

/**
 * @typedef {object} tileCounter - Represent a tile counter.
 * @property {string} modelId - The identifier of the model.
 * @property {TileModel} model - A reference to the tile model.
 * @property {number} count - The number of tiles referenced for this model.
 */

/**
 * @typedef {import('./TileList.js').TileList} TileList
 */

/**
 * @typedef {import('./TileModel.js').TileModel} TileModel
 */

/**
 * Notifies a tile model has been added.
 * @event addmodel
 * @param {string} key - The key of the tile model that was added.
 * @param {TileModel} model - The tile model that was added.
 */

/**
 * Notifies a tile model has been removed.
 * @event removemodel
 * @param {string} key - The key of the tile model that was removed.
 * @param {TileModel} model - The tile model that was removed.
 */

/**
 * Notifies a tile has been added.
 * @event addtile
 * @param {TileModel} tile - The tile that was added.
 */

/**
 * Notifies a tile has been removed.
 * @event removetile
 * @param {TileModel} tile - The tile that was removed.
 */

/**
 * Notifies a value has been set to a counter.
 * @event set
 * @param {string} key - The key of the counter that was set.
 * @param {number} newValue - The new value.
 * @param {number} oldValue - The previous value.
 */

/**
 * Notifies a counter has been removed.
 * @event delete
 * @param {string} key - The key of the counter that was removed.
 * @param {number} value - The last value of the removed counter.
 */

/**
 * Notifies all counters has been removed.
 * @event clear
 */

/**
 * Notifies the counters has been loaded.
 * @event load
 */
