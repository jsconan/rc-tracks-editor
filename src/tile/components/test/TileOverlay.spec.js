/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import { tick } from 'svelte';
import { TileSpecifications } from '../../config';
import { CurvedTileModel, CurvedTileEnlargedModel, StraightTileModel } from '../../models';
import TileOverlay from '../TileOverlay.svelte';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications({ laneWidth, barrierWidth, barrierChunks });
const straightTile = new StraightTileModel(specs);
const curvedTile = new CurvedTileModel(specs);
const curvedTileEnlarged = new CurvedTileEnlargedModel(specs);

describe('TileOverlay', () => {
    it.each([
        [straightTile.type, straightTile],
        [curvedTile.type, curvedTile],
        [curvedTileEnlarged.type, curvedTileEnlarged]
    ])('renders an overlay for a % tile with default values', (type, tile) => {
        const props = {
            tile
        };
        const { container } = render(TileOverlay, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        [straightTile.type, straightTile],
        [curvedTile.type, curvedTile],
        [curvedTileEnlarged.type, curvedTileEnlarged]
    ])('renders an overlay for a % tile with the given parameters', (type, tile) => {
        const props = {
            tile,
            x: 10,
            y: 20,
            angle: 30
        };
        const { container } = render(TileOverlay, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['tile', { tile: curvedTile }],
        ['angle', { angle: 90 }],
        ['x', { x: 40 }],
        ['y', { y: 40 }]
    ])('updates when the parameter %s is modified', async (title, update) => {
        const props = {
            tile: straightTile,
            x: 10,
            y: 20,
            angle: 30
        };
        const rendered = render(TileOverlay, { props });

        await tick();
        rendered.component.$set(update);
        await tick();
        expect(rendered.container).toMatchSnapshot();
    });
});
