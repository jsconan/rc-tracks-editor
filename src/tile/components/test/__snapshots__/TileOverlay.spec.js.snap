// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileOverlay renders an overlay for a % tile with default values 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform=""
    >
      <rect
        fill="var(--color-bg-overlay)"
        height="110"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
        width="90"
        x="-45"
        y="0"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay renders an overlay for a % tile with default values 2`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform=""
    >
      <path
        d="M -45 0 A 10 10 0 0 1 -55 10 L -54.99999999999999 100 A 100 100 0 0 0 45 0 Z"
        fill="var(--color-bg-overlay)"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay renders an overlay for a % tile with default values 3`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform=""
    >
      <path
        d="M -45 0 A 10 10 0 0 1 -55 10 L -55 100 L 0 100 A 45 45 0 0 0 45 55 L 45 0 Z"
        fill="var(--color-bg-overlay)"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay renders an overlay for a % tile with the given parameters 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 10 20)"
    >
      <rect
        fill="var(--color-bg-overlay)"
        height="110"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
        width="90"
        x="-35"
        y="20"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay renders an overlay for a % tile with the given parameters 2`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 10 20)"
    >
      <path
        d="M -35 20 A 10 10 0 0 1 -45 30 L -44.99999999999999 120 A 100 100 0 0 0 55 20 Z"
        fill="var(--color-bg-overlay)"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay renders an overlay for a % tile with the given parameters 3`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 10 20)"
    >
      <path
        d="M -35 20 A 10 10 0 0 1 -45 30 L -45 120 L 10 120 A 45 45 0 0 0 55 75 L 55 20 Z"
        fill="var(--color-bg-overlay)"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay updates when the parameter angle is modified 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(90 10 20)"
    >
      <rect
        fill="var(--color-bg-overlay)"
        height="110"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
        width="90"
        x="-35"
        y="20"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay updates when the parameter tile is modified 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 10 20)"
    >
      <path
        d="M -35 20 A 10 10 0 0 1 -45 30 L -44.99999999999999 120 A 100 100 0 0 0 55 20 Z"
        fill="var(--color-bg-overlay)"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay updates when the parameter x is modified 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 40 20)"
    >
      <rect
        fill="var(--color-bg-overlay)"
        height="110"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
        width="90"
        x="-5"
        y="20"
      />
    </g>
  </div>
</body>
`;

exports[`TileOverlay updates when the parameter y is modified 1`] = `
<body>
  <div>
    <g
      class="tile-overlay"
      transform="rotate(30 10 40)"
    >
      <rect
        fill="var(--color-bg-overlay)"
        height="110"
        stroke="var(--color-fg-overlay)"
        stroke-width="var(--stroke-width-overlay)"
        width="90"
        x="-35"
        y="40"
      />
    </g>
  </div>
</body>
`;
